clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -444,10 +444,20 @@ private boolean maybeSendAndPollTransactionalRequest() {
         AbstractRequest.Builder<?> requestBuilder = nextRequestHandler.requestBuilder();
         Node targetNode = null;
         try {
-            targetNode = awaitNodeReady(nextRequestHandler.coordinatorType());
-            if (targetNode == null) {
+            FindCoordinatorRequest.CoordinatorType coordinatorType = nextRequestHandler.coordinatorType();
+            targetNode = coordinatorType != null ?
+                    transactionManager.coordinator(coordinatorType) :
+                    client.leastLoadedNode(time.milliseconds());
+            if (targetNode != null) {
+                awaitNodeReady(targetNode, coordinatorType);
+            } else if (coordinatorType != null) {
+                log.trace("Coordinator not known for {}, will retry {} after finding coordinator.", coordinatorType, requestBuilder.apiKey());
                 maybeFindCoordinatorAndRetry(nextRequestHandler);
                 return true;
+            } else {
+                log.trace("No nodes available to send requests, polling until a node is ready.");
+                client.poll(retryBackoffMs, time.milliseconds());
+                return true;
```

Comment:
Don't we loose the current `nextRequestHandler` when we end up in this branch?

For instance, if `nextRequestHandler` is a `FindCoordinatorHandler` and there a no nodes available, `targetNode` is `null` and `coordinatorType` is `null` as well so we end up here and poll. We don't do anything with `nextRequestHandler` and return so it is gone. I suppose that it is not an issue for `FindCoordinatorHandler` as a new one will be enqueued automatically when another `TxnRequestHandler` handler is processed and the coordinator is unknown.

We may want to push back `nextRequestHandler` to the queue with `transactionManager.retry(nextRequestHandler)` in oder to handle all the cases.

I am not sure if that could really happen with any other `TxnRequestHandler` type though. What do you think?

Reply:
@rajinisivaram It seems that a non-null `nextRequestHandler` would have been removed from the queue: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java#L872.

I've got the impression that `InitProducerIdHandler` may hit that code path as well as it could have a null `coordinatorType` when the transaction id is not set (idempotent producer). I think that we could unit test this actually by reusing your `testInitProducerIdWithMaxInFlightOne` but without setting the transactional id. Should we add this test to be on the safe side?

Reply:
@dajac Sorry, that was my mistake. Updated code and added test.

Reply:
Thanks. I will make another pass now.